[🐛 Bug]: Next.js 15 App Route Handlers throw TypeError: Illegal invocation: function called with incorrect `this` reference.

### next-on-pages environment related information

```
System:
        Platform: darwin
        Arch: arm64
        Version: Darwin Kernel Version 24.1.0: Thu Oct 10 21:03:15 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T6000
        CPU: (8) arm64 Apple M1 Pro
        Memory: 16 GB
        Shell: /bin/zsh
Package Manager Used: npm (10.7.0)

Relevant Packages:
        @cloudflare/next-on-pages: 1.13.6
        vercel: 39.1.2
        next: 15.0.3
```

### Description

It seems that Next.js in v15.0.0.rc-1 onwards introduced some special handing for `waitUntil` that causes App Route Handlers to error with `Internal Server Error` due to this https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors.
```
✘ [ERROR] TypeError: Illegal invocation: function called with incorrect `this` reference. See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.

      at $.waitUntil
  (file:///Users/juanferreras/Nyla/01_Software/api-test/my-next-app/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/api/hello.func.js:2:826)
      at Ue.handler
  (file:///Users/juanferreras/Nyla/01_Software/api-test/my-next-app/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/api/hello.func.js:2:25694)
      at async p
  (file:///Users/juanferreras/Nyla/01_Software/api-test/my-next-app/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/api/hello.func.js:2:6070)
      at async Module.<anonymous>
  (file:///Users/juanferreras/Nyla/01_Software/api-test/my-next-app/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/api/hello.func.js:8:3795)
```

This happens regardless of using the new `nodejs_compat_v2` (or `nodejs_compat` compatibility flag and compatibility date to `2024-09-23` or later). 

The related source code in Next.js seems to be this:
https://github.com/vercel/next.js/blob/canary/packages/next/src/server/web/spec-extension/fetch-event.ts#L37-L49

### Reproduction

Start a new project using create cloudflare CLI: 
1. `npm create cloudflare@latest -- my-next-app --framework=next` 
2. Update it to Next 15 running `npx @next/codemod@canary upgrade latest` inside `my-next-app` folder

OR clone mine:
1. https://github.com/juanferreras/next-illegal-invocation
2. `npm install`
3. `npm run preview`

### Pages Deployment Method

None

### Pages Deployment ID

N/A

### Additional Information

_No response_

### Would you like to help?

- [X] Would you like to help fixing this bug?

@dario-piotrowicz Hi I am facing the same problem and using the prerelease with the latest vercel CLI does not fix it. But the starter template works for me even with `1.13.6`.

<details>
<summary>logs</summary>

```
 pnpm build:pages

> qiushiyan.dev@0.1.0 build:pages /Users/qiushi/workspace/my-website/qiushiyan.dev
> pnpm next-on-pages

⚡️ @cloudflare/next-on-pages CLI v.0.0.0-28ee1db (prerelease for PR #917)
⚡️ Detected Package Manager: pnpm (9.11.0)
⚡️ Preparing project...
⚡️ Project is ready
⚡️ Building project...
▲  Vercel CLI 39.1.3
▲  Detected `pnpm-lock.yaml` version 9 generated by pnpm@9.x
▲  WARNING: You should not upload the `.next` directory.
▲  Installing dependencies...
▲  Lockfile is up to date, resolution step is skipped
▲  Already up to date
▲  
▲  Done in 1.8s
▲  Detected Next.js version: 15.0.3
▲  Running "pnpm run build"
▲  > qiushiyan.dev@0.1.0 build /Users/qiushi/workspace/my-website/qiushiyan.dev
▲  > next build

... other outputs

▲  Collecting page data ...
▲  ⚠ Using edge runtime on a page currently disables static generation for that page
▲  Generating static pages (0/10) ...
▲  Generating static pages (2/10)
▲  Generating static pages (4/10) 
▲  Generating static pages (7/10)
▲  ✓ Generating static pages (10/10)
▲  Finalizing page optimization ...
▲  Collecting build traces ...
▲  
▲  Route (app)                                                    Size     First Load JS
▲  ┌ ƒ /                                                          6.53 kB         168 kB
▲  ├ ○ /_not-found                                                900 B           101 kB
▲  ├ ○ /about                                                     3.75 kB         135 kB
▲  ├ ƒ /api/og                                                    146 B           100 kB
▲  ├ ○ /notes                                                     1.3 kB          200 kB
▲  ├ ƒ /notes/[slug]                                              3.75 kB         148 kB
▲  ├ ƒ /posts                                                     3.66 kB         223 kB
▲  ├ ƒ /posts/[slug]                                              5.3 kB          146 kB
▲  ├ ○ /projects/qlang                                            2 kB            126 kB
▲  ├ ○ /recipes                                                   1.3 kB          200 kB
▲  └ ● /recipes/[group]/[slug]                                    151 kB          275 kB
▲    ├ /recipes/python/polymorphism-over-if-else
▲    └ /recipes/python/creating-data-pipelines-with-generators
▲  + First Load JS shared by all                                  100 kB
▲  ├ chunks/1102d547-4f7e829b20382639.js                        52.5 kB
▲  ├ chunks/2693-77129af5078baaf0.js                            45.4 kB
▲  └ other shared chunks (total)                                2.29 kB
▲  ○  (Static)   prerendered as static content
▲  ●  (SSG)      prerendered as static HTML (uses generateStaticParams)
▲  ƒ  (Dynamic)  server-rendered on demand
▲  Traced Next.js server files in: 124.74ms
▲  Created all serverless functions in: 98.619ms
▲  Collected static files (public/, static/, .next/static): 3.714ms
▲  Build Completed in .vercel/output [40s]
⚡️ Completed `pnpm exec vercel build`.
⚡️ Invalid prerender config for /recipes/[group]/[slug]
⚡️ Invalid prerender config for /recipes/[group]/[slug].rsc

⚡️ Build Summary (@cloudflare/next-on-pages v0.0.0-28ee1db (prerelease for PR #917))
⚡️ 
⚡️ Edge Function Routes (5)
⚡️   ┌ /
⚡️   ├ /api/og
⚡️   ├ /notes/[slug]
⚡️   ├ /posts
⚡️   └ /posts/[slug]
⚡️ 
⚡️ Prerendered Routes (13)
⚡️   ┌ /about
⚡️   ├ /about.rsc
⚡️   ├ /favicon.ico
⚡️   ├ /notes
⚡️   ├ /notes.rsc
⚡️   ├ /projects/qlang
⚡️   ├ /projects/qlang.rsc
⚡️   ├ /recipes
⚡️   ├ /recipes.rsc
⚡️   ├ /recipes/python/creating-data-pipelines-with-generators
⚡️   ├ /recipes/python/creating-data-pipelines-with-generators.rsc
⚡️   ├ /recipes/python/polymorphism-over-if-else
⚡️   └ /recipes/python/polymorphism-over-if-else.rsc
⚡️ 
⚡️ Wasm Files (2)
⚡️   ┌ wasm_77d9faebf7af9e421806970ce10a58e9d83116d7
⚡️   └ wasm_ef4866ecae192fd87727067cf2c0c0cf9fb8b020
⚡️ 
⚡️ Bundled Assets (2)
⚡️   ┌ SpaceGrotesk-SemiBold.492effd4b68e572b.ttf
⚡️   └ noto-sans-v27-latin-regular.5dda3fca77107598.ttf
⚡️ 
⚡️ Other Static Assets (81)
⚡️   ┌ /_app.rsc.json
⚡️   ├ /_document.rsc.json
⚡️   ├ /_error.rsc.json
⚡️   ├ /.DS_Store
⚡️   └ ... 77 more

⚡️ Build log saved to '.vercel/output/static/_worker.js/nop-build-log.json'
⚡️ Generated '.vercel/output/static/_worker.js/index.js'.
⚡️ Build completed in 1.44s


❯ wrangler pages dev

 ⛅️ wrangler 3.93.0
-------------------

Attaching additional modules:
┌────────────────────────────────────────────────────────────────────────────────────┬───────────────┬─────────────┐
│ Name                                                                               │ Type          │ Size        │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/assets/SpaceGrotesk-SemiBold.492effd4b68e572b.ttf.bin       │ buffer        │ 84.55 KiB   │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/assets/noto-sans-v27-latin-regular.5dda3fca77107598.ttf.bin │ buffer        │ 27.10 KiB   │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/cache/adaptor.js                                            │ esm           │ 2.29 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/cache/cache-api.js                                          │ esm           │ 0.43 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/cache/kv.js                                                 │ esm           │ 0.35 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/functions/api/og.func.js                                    │ esm           │ 505.91 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/functions/index.func.js                                     │ esm           │ 321.61 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/functions/notes/[slug].func.js                              │ esm           │ 138.87 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/functions/posts/[slug].func.js                              │ esm           │ 179.26 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/functions/posts.func.js                                     │ esm           │ 125.67 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/manifest/88a389dd6c068190adb2efc795e4063c.js                │ esm           │ 1.89 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/wasm/wasm_77d9faebf7af9e421806970ce10a58e9d83116d7.wasm     │ compiled-wasm │ 1346.05 KiB │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/wasm/wasm_ef4866ecae192fd87727067cf2c0c0cf9fb8b020.wasm     │ compiled-wasm │ 86.58 KiB   │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/11881.js                                            │ esm           │ 0.23 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/16331.js                                            │ esm           │ 0.45 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/16618.js                                            │ esm           │ 4.04 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/25c51a0a3a0853d7873f6afbad602fa6.js                 │ esm           │ 227.06 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/36728795507221d3b28b2712a7a9760a.js                 │ esm           │ 1.94 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/83053.js                                            │ esm           │ 0.18 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/93a5f1e4bfbf6611f36876adf2fa1bf5.js                 │ esm           │ 143.71 KiB  │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/9535.js                                             │ esm           │ 0.23 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/a218bd64825a354e5323ddaedd1c493c.js                 │ esm           │ 0.86 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/a6a8a1a4cb307f3b873a94c8ec93a56c.js                 │ esm           │ 7.11 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/d5e43cbf381cde3e9d2e371baaa5876a.js                 │ esm           │ 3.56 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/f430aae95b006d021e2a11b626956392.js                 │ esm           │ 0.10 KiB    │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ __next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js                 │ esm           │ 1893.33 KiB │
├────────────────────────────────────────────────────────────────────────────────────┼───────────────┼─────────────┤
│ Total (26 modules)                                                                 │               │ 5103.34 KiB │
└────────────────────────────────────────────────────────────────────────────────────┴───────────────┴─────────────┘
✨ Compiled Worker successfully
✨ Parsed 1 valid header rule.
Using vars defined in .dev.vars
Your worker has access to the following bindings:
- D1 Databases:
  - DB: personal-website (DB), Preview: (DB) (local)
⎔ Starting local server...
[wrangler:inf] Ready on http://localhost:8788
✘ [ERROR] ⨯ TypeError: Illegal invocation: function called with incorrect `this` reference. See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.

      at null.<anonymous>
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:13:16764)
      at Pa.dA.__chunk_5711
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:13:16870)
      at u
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/index.func.js:1:39624)
      at Pa.dA.__chunk_18357
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:1:17257)
      at u
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/index.func.js:1:39624)
      at h.__chunk_88847
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/a6a8a1a4cb307f3b873a94c8ec93a56c.js:1:1879)
      at u
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/index.func.js:1:39624)
      at h.__chunk_16618
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/16618.js:1:1380)
      at u
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/index.func.js:1:39624)
      at 19789
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/functions/index.func.js:23:8572)


✘ [ERROR] ⨯ TypeError: Cannot read properties of null (reading 'default')

      at Object.loadComponent
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:106:10544)
      at Gn.findPageComponents
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:106:5898)
      at Gn.renderErrorToResponseImpl
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:93:11447)
      at async Gn.renderToResponseImpl
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:93:9528)
      at async Gn.pipeImpl
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:91:17503)
      at async Gn.handleCatchallRenderRequest
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:106:3301)
      at async Gn.runImpl
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:91:17224)
      at async Gn.handleRequestImpl
  (file:///Users/qiushi/workspace/my-website/qiushiyan.dev/.vercel/output/static/_worker.js/__next-on-pages-dist__/webpack/f5dc104510594ceb6f204bc82384b10b.js:91:16050)


[wrangler:inf] GET / 500 Internal Server Error (128ms)
```
</details>

--- a/.changeset/bright-vans-relate.md
+++ b/.changeset/bright-vans-relate.md
@@ -1,0 +1,5 @@
+---
+'@cloudflare/next-on-pages': patch
+---
+
+fix Next.js calling a non-bound waitUntil method

--- a/.changeset/bright-vans-relate.md
+++ b/.changeset/bright-vans-relate.md
@@ -2,4 +2,4 @@
 '@cloudflare/next-on-pages': patch
 ---
 
-fix Next.js calling a non-bound waitUntil method
+fix Next.js calling a non-bound waitUntil function

--- a/packages/next-on-pages/src/buildApplication/processVercelFunctions/dedupeEdgeFunctions.ts
+++ b/packages/next-on-pages/src/buildApplication/processVercelFunctions/dedupeEdgeFunctions.ts
@@ -635,8 +635,8 @@ function fixFunctionContents(contents: string): string {
 	// We need to update it so that uses a `waitUntil` bound to our `ctx` object
 	// (source https://github.com/vercel/next.js/blob/b7c271d7/packages/next/src/server/web/spec-extension/fetch-event.ts#L21)
 	contents = contents.replace(
-		/this\[(.)\]=(\w)\?{kind:"external",function:\2}:{kind:"internal",promises:\[(.*?)\]}/gm,
-		`this[$1]=e?{kind:"external",function:globalThis[Symbol.for('__cloudflare-request-context__')].ctx.waitUntil.bind(globalThis[Symbol.for('__cloudflare-request-context__')].ctx)}:{kind:"internal",promises:[$3]}`,
+		/this\[(.)\]=(\w)\?{kind:"external",function:\2}:{kind:"internal",promises:/gm,
+		`this[$1]=$2?{kind:"external",function:globalThis[Symbol.for('__cloudflare-request-context__')].ctx.waitUntil.bind(globalThis[Symbol.for('__cloudflare-request-context__')].ctx)}:{kind:"internal",promises:`,
 	);
 
 	return contents;

--- a/packages/next-on-pages/src/buildApplication/processVercelFunctions/dedupeEdgeFunctions.ts
+++ b/packages/next-on-pages/src/buildApplication/processVercelFunctions/dedupeEdgeFunctions.ts
@@ -631,6 +631,14 @@ function fixFunctionContents(contents: string): string {
 		'const { originalRequest = $1 } = $1$3',
 	);
 
+	// Next.js calls `waitUntil` without binding is to the context object, triggering the following error: `TypeError: Illegal invocation: function called with incorrect `this` reference`
+	// We need to update it so that uses a `waitUntil` bound to our `ctx` object
+	// (source https://github.com/vercel/next.js/blob/b7c271d7/packages/next/src/server/web/spec-extension/fetch-event.ts#L21)
+	contents = contents.replace(
+		/this\[(.)\]=(\w)\?{kind:"external",function:\2}:{kind:"internal",promises:\[(.*?)\]}/gm,
+		`this[$1]=e?{kind:"external",function:globalThis[Symbol.for('__cloudflare-request-context__')].ctx.waitUntil.bind(globalThis[Symbol.for('__cloudflare-request-context__')].ctx)}:{kind:"internal",promises:[$3]}`,
+	);
+
 	return contents;
 }
 

